### ISSUE TITLE
Feature Request: findAll() Support Lucee's ReturnType/ColumnKey

### ISSUE THREAD

**Issue by geirman** [wheels-dev/wheels#945]
## Feature Request
As of Lucee 5.x, cfquery allows the developer to define various returntypes (e.g. query, array_of_entity, array, struct). Currently CFWheels has `findAll(returnAs="")` which is similar. A value of `object` is similar to Lucee's `returnType="array"`. It would be really cool if CFW supported all the return types, especially `struct`. Unfortunately, ACF doesn't have any similar functionality so they'd probably require a UDF shim of some sort. 

## Use Case 
I have a table of quotes. Some quotes are potential jobs, some of which will be converted to jobs, but not all. The jointype on the jobs table would be `joinType=outer` with a FK related to the quoteId. 

When querying all quotes, it would be awesome if we could also query `jobs=model("jobs").findAll(where="eventId IS NOT NULL", returnAs="struct", columnKey="quoteId")`

Then, as I'm looping through my `quotes` query, I could check `jobs[quoteId]` for related event data. 

## Reference
https://docs.lucee.org/reference/tags/query.html
https://helpx.adobe.com/coldfusion/cfml-reference/coldfusion-tags/tags-p-q/cfquery.html

**Comment by perdjurner** [comment]
Isn't this already supported?

From the docs:

> Set to objects to return an array of objects, set to structs to return an array of structs, or set to query to return a query result set.

**Comment by geirman** [comment]
Hey @perdjurner, I was hoping, but unfortunately they're different. I've created a demo page that shows all the various outputs so you can compare, but here are the highlights. Notice that `Lucee as Struct` outputs an actual struct where the struct's key is the object's id (defined as `columnKey`) whereas CFW returns an array.

![2019-05-10_0838 | CFW as Struct](https://user-images.githubusercontent.com/1640318/57539360-30a72c80-72ff-11e9-801a-a85f0efeb1e1.png)
![2019-05-10_0837 | Lucee as Struct](https://user-images.githubusercontent.com/1640318/57539370-36047700-72ff-11e9-85c2-90382089eccd.png)
 
# All Examples
![2019-05-10_0839 | All](https://user-images.githubusercontent.com/1640318/57539596-b0cd9200-72ff-11e9-9373-8eb999049051.png)

**Comment by perdjurner** [comment]
I see, yes, sounds like it would be a good enhancement then :)

**Comment by andybellenie** [comment]
I agree that it would be clearer to be returnAs="array" rather than "structs". I'd be happy to change this and deprecate the old usage.

Is there a real world reason why the Lucee format might be preferable?

**Comment by andybellenie** [comment]
Sorry I didn't fully read your original issue regarding using a custom key column. I still don't quite follow what it is you're trying to do, though.

**Comment by geirman** [comment]
@andybellenie Sorry, I'll try to explain it better. I have a quotes table. As I output each quote, I'll want to add a tag to the one's which have converted to an actual job, which will be a small percentage. Actually, in my case, they become an "event", but I thought job would be a better generic reference for the use case above. In any case, as I loop through each quote to build the table, I might want to do something like this...

```
<cfset events = model("events").findAll(where="quoteId IS NOT NULL", returnAs="struct", columnKey="quoteId">
<cfloop query="#quotes#">
 <td>
   #fullName#
   <cfif StructKeyExists(events, id)>
     <!--- output event related data --->
   <cfelse>
     <!--- tag as "unsold" --->
   </cfif>
 </td>
</cfloop>
```

**Comment by andybellenie** [comment]
I don't understand your code. The 'quotes' variable doesn't exist. If you meant 'events' then it isn't a query.

Based on your description, can't you left join quote to jobs and only display job related fields if IsNumeric(quotes.jobid) is true?

**Comment by geirman** [comment]
Sorry, I updated the code example. There's definitely other ways to go about this.

**Comment by andybellenie** [comment]
Thanks, that makes more sense now :)

I'd like to see a real world situation where there isn't an existing solution before I'd support this change.

**Comment by andybellenie** [comment]
At least, where the existing solution isn't equally effective and clean, code wise.

**Comment by geirman** [comment]
Well, I'm not sure there is an existing solution that is clean. In my example, I want to show all quotes, but add job related data when that data exists (which would be a small percentage of time, about 10% of them). If I understand your suggestion correctly, you're limiting quotes to just those which have job related data. 

Without this, I'd probably just `returnAs="array"` then map that array to create the struct in the desired form. Definitely not a big deal, but since Lucee bakes this right in, then I thought it was worth considering.

**Comment by geirman** [comment]
Ben Nadel mentioned this feature in his recent blog post. I thought it might be interesting to share how he's using it...
https://www.bennadel.com/blog/3654-returning-cfquery-query-results-as-arrays-in-lucee-5-3-2-77.htm

**Comment by zainforbjs** [comment]
@geirman @andybellenie Does this still need to be done?

**Comment by geirman** [comment]
@zainforbjs doesn't sound like @andybellenie supports it

**Comment by chapmandu** [comment]
returnType is now supported.

**Comment by zainforbjs** [comment]
@geirman The return as has been enhanced as per the issue, Please have your review under v3.0 and give us your feedback. Thank you for your support.
